rust/src/mqtt/detect.rs

```diff
@@ -0,0 +1,558 @@
+            if let Some(p) = &cv.will_message {
+                if p.len() > 0 {
+                    unsafe {
+                        *buffer = p.as_ptr();
+                        *buffer_len = p.len() as u32;
+                    }
+                    return 1;
+                }
+            }
+        }
+    }
+
+    unsafe {
+        *buffer = ptr::null();
+        *buffer_len = 0;
+    }
+
+    return 0;
+}
+
+#[no_mangle]
+pub extern "C" fn rs_mqtt_tx_get_connack_sessionpresent(
+    tx: &MQTTTransaction,
+    session_present: *mut bool,
+) -> u8 {
+    for msg in tx.msg.iter() {
+        if let MQTTOperation::CONNACK(ref ca) = msg.op {
+            unsafe {
+                *session_present = ca.session_present;
+            }
+            return 1;
+        }
+    }
+    return 0;
+}
+
+#[no_mangle]
+pub extern "C" fn rs_mqtt_tx_get_publish_topic(
+    tx: &MQTTTransaction,
+    buffer: *mut *const u8,
+    buffer_len: *mut u32,
+) -> u8 {
+    for msg in tx.msg.iter() {
+        if let MQTTOperation::PUBLISH(ref pubv) = msg.op {
+            let p = &pubv.topic;
+            if p.len() > 0 {
+                unsafe {
+                    *buffer = p.as_ptr();
+                    *buffer_len = p.len() as u32;
+                }
+                return 1;
+            }
+        }
+    }
+
+    unsafe {
+        *buffer = ptr::null();
+        *buffer_len = 0;
+    }
+
+    return 0;
+}
+
+#[no_mangle]
+pub extern "C" fn rs_mqtt_tx_get_publish_message(
+    tx: &MQTTTransaction,
+    buffer: *mut *const u8,
+    buffer_len: *mut u32,
+) -> u8 {
+    for msg in tx.msg.iter() {
+        if let MQTTOperation::PUBLISH(ref pubv) = msg.op {
+            let p = &pubv.message;
+            if p.len() > 0 {
+                unsafe {
+                    *buffer = p.as_ptr();
+                    *buffer_len = p.len() as u32;
+                }
+                return 1;
+            }
+        }
+    }
+
+    unsafe {
+        *buffer = ptr::null();
+        *buffer_len = 0;
+    }
+
+    return 0;
+}
+
+#[no_mangle]
+pub extern "C" fn rs_mqtt_tx_get_topic(tx: &MQTTTransaction,
+                                       i: u16,
+                                       buf: *mut *const u8,
+                                       len: *mut u32)
+                                       -> u8
+{
+    for msg in tx.msg.iter() {
+        match msg.op {
+            MQTTOperation::SUBSCRIBE(ref subv) => {
+                if (i as usize) < subv.topics.len() {
+                    let topic = &subv.topics[i as usize];
+                    if topic.topic_name.len() > 0 {
+                        unsafe {
+                            *len = topic.topic_name.len() as u32;
+                            *buf = topic.topic_name.as_ptr();
+                        }
+                        return 1;
+                    }
+                }
+            }
+            MQTTOperation::UNSUBSCRIBE(ref unsubv) => {
```

Comment:
No, there's no `SUBSCRIBE`-`UNSUBSCRIBE` in the same TX. The fact that this function checks both is just to educe code duplication because both have multiple topics that need to be accessed via index from Rust. IMHO I don't want to do semantic checks in the detection code: I want to get the data out of the TX, if is in there. So, if someone wants to match a topic (via the `mqtt.topic` sticky buffer) I don't care in the scope of this keyword in what message appears, I want to flag the TX as being relevant if any topic in any message type in there matches.
Further restrictions used in the signature (e.g. whether the topic should be in `SUBSCRIBE`, `UNSUBSCRIBE`, etc.) should be handled by another keyword (`mqtt.type`).
I'll add a clarifying comment.

Reply:
Thanks for the explanation